cvmfs/cache_ram.h

```diff
@@ -13,75 +13,275 @@
+      n_readahead = statistics->Register(name + ".n_readahead",
+        "Number of ReadAhead calls for " + name);
+      n_sizeoftxn = statistics->Register(name + ".n_sizeoftxn",
+        "Number of SizeOfTxn calls for " + name);
+      n_starttxn = statistics->Register(name + ".n_starttxn",
+        "Number of StartTxn calls for " + name);
+      n_ctrltxn = statistics->Register(name + ".n_ctrltxn",
+        "Number of CtrlTxn calls for " + name);
+      n_write = statistics->Register(name + ".n_write",
+        "Number of Write calls for " + name);
+      n_reset = statistics->Register(name + ".n_reset",
+        "Number of Reset calls for " + name);
+      n_openfromtxn = statistics->Register(name + ".n_openfromtxn",
+        "Number of OpenFromTxn calls for " + name);
+      n_aborttxn = statistics->Register(name + ".n_aborttxn",
+        "Number of AbortTxn calls for " + name);
+      n_committxn = statistics->Register(name + ".n_committxn",
+        "Number of Commit calls for " + name);
+      n_enfile = statistics->Register(name + ".n_enfile",
+        "Number of times " + name + " reached the limit on handles");
+      n_openregular = statistics->Register(name + ".n_openregular",
+        "Number of opens from the regular cache for " + name);
+      n_openvolatile = statistics->Register(name + ".n_openvolatile",
+        "Number of opens from the volatile cache for " + name);
+      n_openmiss = statistics->Register(name + ".n_openmiss",
+        "Number of missed opens for " + name);
+      n_realloc = statistics->Register(name + ".n_realloc",
+        "Number of reallocs for " + name);
+      n_overrun = statistics->Register(name + ".n_overrun",
+        "Number of cache limit overruns for " + name);
+      n_full = statistics->Register(name + ".n_full",
+        "Number of overruns that could not be resolved for " + name);
+    }
+  };
+
   virtual CacheManagerIds id() { return kRamCacheManager; }
 
-  RamCacheManager();
-  virtual ~RamCacheManager();
+  RamCacheManager(
+    uint64_t max_size,
+    unsigned max_entries,
+    perf::Statistics *statistics)
+    : max_size_(max_size)
+    , regular_entries_(max_entries/2, "RamCache.regular", statistics)
+    , volatile_entries_(max_entries/2, "RamCache.volatile", statistics)
+    , counters_(statistics, "RamCache") {
+    int retval = pthread_rwlock_init(&rwlock_, NULL);
+    assert(retval == 0);
+    LogCvmfs(kLogCache, kLogDebug, "max %u B, %u entries",
+             max_size, max_entries);
+  }
+  virtual ~RamCacheManager() {
+    pthread_rwlock_destroy(&rwlock_);
+  }
   virtual bool AcquireQuotaManager(QuotaManager *quota_mgr);
 
+  /**
+   * Open a new file descriptor into the cache. Note that opening entries effectively pins them in the cache,
+   * so it may be necessary to close unneeded file descriptors to allow eviction to make room in the cache
+   * @param id The hash key
+   * @returns The smallest free integer file descriptor, or -ENOENT if the entry
```

Comment:
We might soften the requirement on the return value to "an integer between 0 and max number of open fds".